src/twister2/device/native_simulator_adapter.py

```diff
@@ -23,78 +29,105 @@ def __init__(self, twister_config, hardware_map: HardwareMap | None = None, **kw
+        end_time = time.time() + timeout
+        while not self._stop_job and not self._process.stdout.at_eof():
+            try:
+                line = await asyncio.wait_for(self._process.stdout.readline(), timeout=0.1)
+            except asyncio.TimeoutError:
+                pass
             else:
-                if stderr := self._process.stderr.read():
-                    logger.error(stderr.decode())
-                raise TwisterFlashException(f'Could not run simulator with PID {self._process.pid}')
-
-    def _collect_process_output(self, process: subprocess.Popen) -> threading.Thread:
-        """Create Thread which saves a process output to a file."""
-        def _read():
-            with process.stdout:
-                for line in iter(process.stdout.readline, b''):
-                    self.queue.put(line.decode().strip())
-
-        return threading.Thread(target=_read, daemon=True)
-
-    def _wait_and_terminate_process(self, process: subprocess.Popen, timeout: float) -> threading.Thread:
-        """Create Thread which kills a process after given time."""
-        def _waiting():
-            end_time = time.time() + timeout
-            while process.poll() is None and time.time() < end_time:
-                time.sleep(0.1)
-            if process.poll() is None:
-                process.kill()
-                logger.error('Process %s terminated after %s seconds', process.pid, timeout)
+                if line:
+                    self.queue.put(line.decode('utf-8').strip())
+            if time.time() > end_time:
                 self._process_ended_with_timeout = True
-        return threading.Thread(target=_waiting, daemon=True)
+                logger.debug(f'Finished process with PID {self._process.pid} after {timeout} seconds timeout')
+                break
+
+        self.queue.put(END_DATA)
+        return await self._process.wait()
+
+    def _task(self) -> None:
+        command: list[str] = self._get_command(self.build_dir)
+        logger.info('Running command: %s', command)
+        try:
+            return_code: int = asyncio.run(
+                self._run_command(command, timeout=self.timeout)
+            )
+        except subprocess.SubprocessError as e:
+            logger.error('Flashing failed due to subprocess error %s', e)
+            self._exc = TwisterFlashException(e.args)
+        except FileNotFoundError as e:
+            logger.error(f'Flashing failed due to file not found: {e.filename}')
+            self._exc = TwisterFlashException(f'File not found: {e.filename}')
+        except Exception as e:
+            logger.error('Flashing failed: %s', e)
+            self._exc = TwisterFlashException(e.args)
+        else:
+            if return_code == 0:
+                logger.info('Flashing finished with success')
+            elif return_code > 0:
+                self._exc = TwisterFlashException(f'Flashing finished with errors for PID {self._process.pid}')
```

Comment:
I wondering if raising error every time when return code is not equal to 0 is a good idea. ZTest tests are designed to return -1 when some of test inside failed. With such proposed workflow such test will be marked as failed (by log parser) and then additionally as an error (in dut fixture postcondition). In ZTest case test should be marked only once as a failed with no additional information about error. So, I propose to add to `DeviceAbstract` class a flag like `ignore_return_code` (default value `False`) and then in `parser` fixture when `ztest` parser is created this flag can be changed into `True`. Then such a flag can be used here to decide if return code error should be raised or not.

Reply:
Changed. Now we log only return_code without raising an exception.